magento/pwa-studio
# [PWA-355]: Add richer extension points and documentation

## Description

Add features to the extensibility framework to support rich integration with Peregrine talons. This enables extensions that can deeply integrate analytics and segmentation code.

### Features
- Changed Buildpack `wrapEsmModules` target to a lower-level, more versatile `transformModules` target
  - Original module-wrap functionality is still available, using a transform of type `source` and with transformModule `wrap-esm-loader`.
  - New transform type `babel` allows for AST manipulation.
  - Interceptors of `transformModules` can only intercept their own files.
- Added Peregrine `talons` target for wrapping talon functions
  - Only `App.useApp` and `ProductFullDetail.useProductFullDetail` implemented so far
  - Infinitely composable
- Added VeniaUI `routes` target for adding custom routes
  - Refactored Venia's existing custom routes to dogfood this target
- Added a suite of test helpers for developers making their own extensions
  - Mock builds and build buses for simulating a full build with codemods
  - Mock targets for testing intercept behavior
  - Mock Webpack loaders for testing source code transforms

### Changes
- Refactored `MagentoResolver` to support some common development scenarios
  - Symlinked packages
  - Packages referencing each other from targets
  - Resolving a file path using Webpack rules, before running the Webpack compiler
  - Legacy behavior still supported
- Refactored `configureWebpack` into smaller modules for creating different parts of Webpack config
  - Test helpers can now simulate a closer-to-real Webpack config by composing these together
  - transformModules target implementation now encapsulated in JS module rules
- Upgraded Jest to v25
  - Performance and snapshot reliability were taking a dive with the extra tests
  - Changed some harnesses to be compatible with v25
  - Changed some snapshots to be more tolerant of versioning

### Docs
- Refactored several Buildpack modules to be more readable structures
- Put JSDoc comments on everything
  - Properly formatted JSDoc comments will enable intellisense in VSCode and IntelliJ environments
  - @jcalcaben may also be able to generate formatted documentation from the JSDoc comments
- Wrote an introductory developer and user tutorial for Targets


## Related Issue
Closes #PWA-355.

## Acceptance 

### Verification Stakeholders
@magento/pwa-studio-team 


### Verification Steps
There is a step-by-step tutorial in the getting started guide! This can serve as a verification repro.
[Quick Start](https://github.com/magento/pwa-studio/blob/zetlen/talon-targets/pwa-devdocs/_drafts/pwa-studio-fundamentals/extensibility-with-targets/index.md#quick-start)

Proceed through the steps and verify that you can create a new custom route from within the repository.

Additional verification: create a separate package to implement the same functionality as a module. See [Extension Development](https://github.com/magento/pwa-studio/blob/zetlen/talon-targets/pwa-devdocs/_drafts/pwa-studio-fundamentals/extensibility-with-targets/index.md#extension-development)


## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
